app/src/org/commcare/CommCareHeartbeatManager.java

```diff
@@ -0,0 +1,309 @@
+
+        @Override
+        public void processSuccess(int responseCode, InputStream responseData) {
+            try {
+                String responseAsString = StreamsUtil.inputStreamToByteArray(responseData).toString();
+                JSONObject jsonResponse = new JSONObject(responseAsString);
+                parseHeartbeatResponse(jsonResponse);
+            }
+            catch (JSONException e) {
+                Logger.log(AndroidLogger.TYPE_ERROR_SERVER_COMMS,
+                        "Heartbeat response was not properly-formed JSON: " + e.getMessage());
+            }
+            catch (IOException e) {
+                Logger.log(AndroidLogger.TYPE_ERROR_SERVER_COMMS,
+                        "IO error while processing heartbeat response: " + e.getMessage());
+            }
+        }
+
+        @Override
+        public void processRedirection(int responseCode) {
+            processErrorResponse(responseCode);
+        }
+
+        @Override
+        public void processClientError(int responseCode) {
+            processErrorResponse(responseCode);
+        }
+
+        @Override
+        public void processServerError(int responseCode) {
+            processErrorResponse(responseCode);
+        }
+
+        @Override
+        public void processOther(int responseCode) {
+            processErrorResponse(responseCode);
+        }
+
+        @Override
+        public void handleIOException(IOException exception) {
+            Logger.log(AndroidLogger.TYPE_ERROR_SERVER_COMMS,
+                    "Encountered IOException while getting response stream for heartbeat response: "
+                            + exception.getMessage());
+        }
+
+        private void processErrorResponse(int responseCode) {
+            Logger.log(AndroidLogger.TYPE_ERROR_SERVER_COMMS,
+                    "Received error response from heartbeat request: " + responseCode);
+        }
+    };
+
+    private static CommCareHeartbeatManager INSTANCE;
+    public static CommCareHeartbeatManager instance() {
+        if (INSTANCE == null) {
+            INSTANCE = new CommCareHeartbeatManager();
+        }
+        return INSTANCE;
+    }
+
+    public void startHeartbeatCommunications() {
+        TimerTask heartbeatTimerTask = new TimerTask() {
+            @Override
+            public void run() {
+                try {
+                    User currentUser = CommCareApplication.instance().getSession().getLoggedInUser();
+                    //simulateRequestGettingStuck();
+                    //requestHeartbeat(currentUser);
+                    parseTestHeartbeatResponse();
+                } catch (SessionUnavailableException e) {
+                    // Means the session has ended, so we should stop these requests
+                    stopHeartbeatCommunications();
+                }
+            }
+        };
+
+        heartbeatTimer = new Timer();
+        heartbeatTimer.schedule(heartbeatTimerTask, new Date(), ONE_DAY_IN_MS);
```

Comment:
The scheduling doesn't quite match my expectations. based on what we'd discussed previously, I think I would expect this process to be

[Manager] -> [One Thread Ever]

Manager starts the thread and wraps its events/lifecycle.

one- thread-ever is in a while() loop, that exits If and When [The thread succeeds] OR it reads a "stopPolling" flag 

stopHeartBeat -> one-thread-ever.stopPolling = true;

at the end of the while loop, if I/O was unsuccesful, the thead sleeps for a while(60 seconds)

When the thread wakes up it checks a flag that says "you should try again right now", if that's set to true, it tries again, if not it keeps sleeping in 60 second increments until some longer backoff (minutes/hours)

When a form push succeeds,the app can say "session.tellTheHeartbeatToTry->Manager->the-one-thread.tryagainflag = true"